app/models/targeting.rb

```diff
@@ -0,0 +1,28 @@
+class Targeting < ActiveRecord::Base
+
+  belongs_to :user
+  belongs_to :bookmark
+
+  has_many :targeting_hosts, :dependent => :destroy
+  has_many :hosts, :through => :targeting_hosts
+  has_many :template_invocations, :dependent => :destroy
+
+  validates :bookmark_id, :presence => {:message => N_("must be specified if search query is blank")},
+            :if => Proc.new { |a| a.search_query.blank? }
+
+  validate :bookmark_or_query
```

Comment:
a little be overlapping with the above validation? maybe unify it in one custom validation?